File: gamelogic/src/main/java/com/alta_v2/game/gamelogic/domain/npc/RepeatableActionProcessorImpl.java
package com.alta_v2.game.gamelogic.domain.npc;

import com.alta_v2.facade.tiledMapApi.TiledMapCoreApi;
import com.alta_v2.game.gamelogic.data.npc.NpcModel;
import com.alta_v2.game.gamelogic.utils.LogicThreadFactory;
import com.alta_v2.physics.task.MovementDirection;
import com.alta_v2.physics.task.resultObserver.TaskResultObserver;
import com.google.common.collect.Lists;
import lombok.RequiredArgsConstructor;
import lombok.extern.log4j.Log4j2;

import javax.inject.Inject;
import java.util.List;
import java.util.concurrent.Executors;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.TimeUnit;

@Log4j2
@RequiredArgsConstructor(onConstructor = @__(@Inject))
public class RepeatableActionProcessorImpl implements RepeatableActionProcessor {

    private static final String LISTENER_THREAD_NAME = "repeatable-action-processing";

    private final List<NpcModel> npcList = Lists.newCopyOnWriteArrayList();
    private final ScheduledExecutorService executorService = Executors.newScheduledThreadPool(
            1, new LogicThreadFactory(LISTENER_THREAD_NAME)
    );
    private final TiledMapCoreApi tiledMapCoreApi;

    /**
     * {@inheritDoc}
     */
    @Override
    public void addToRandomMovement(NpcModel npc) {
        npc.setLastMovementMills(System.currentTimeMillis());
        this.npcList.add(npc);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void startAsync() {
        this.executorService.scheduleWithFixedDelay(this::doNpcMovement, 0L, 100L, TimeUnit.MILLISECONDS);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void destroy() {
        try {
            this.executorService.shutdown();
        } catch (Exception e) {
            log.error(e);
        }
        this.npcList.clear();
    }

    private void doNpcMovement() {
        this.npcList.forEach(npc -> {
            if (npc.isMovementRunning()) {
                return;
            }

            if (System.currentTimeMillis() - npc.getLastMovementMills() < npc.getRepeatMovementInterval()) {
                return;
            }

            TaskResultObserver observer = tiledMapCoreApi.performNpcMovement(npc.getId(), MovementDirection.randomDirection());
            if (observer != null) {
                npc.setMovementRunning(true);
                observer.subscribeOnComplete(() -> {
                    npc.setMovementRunning(false);
                    npc.setLastMovementMills(System.currentTimeMillis());
                });
            }
        });
    }
}

File: gamelogic/src/main/kotlin/com/alta_v2/game/gamelogic/domain/npc/RepeatableActionProcessorImpl.kt
package com.alta_v2.game.gamelogic.domain.npc

import com.alta_v2.facade.tiledMapApi.TiledMapCoreApi
import com.alta_v2.game.gamelogic.data.npc.NpcModel
import com.alta_v2.game.gamelogic.utils.LogicThreadFactory
import com.alta_v2.physics.task.MovementDirection
import com.google.inject.Inject
import mu.KotlinLogging
import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.function.Consumer

private const val LISTENER_THREAD_NAME = "repeatable-action-processing"
private val log = KotlinLogging.logger {  }

class RepeatableActionProcessorImpl @Inject constructor(val tiledMapCoreApi: TiledMapCoreApi) : RepeatableActionProcessor {

    private val npcList = CopyOnWriteArrayList<NpcModel>()
    private val npcMovementConsumer = Consumer { npc: NpcModel -> this.performNpcMovement(npc) }

    private val executorService = Executors.newScheduledThreadPool(
            1, LogicThreadFactory(LISTENER_THREAD_NAME)
    )

    override fun addToRandomMovement(npc: NpcModel) {
        npc.lastMovementMills = System.currentTimeMillis()
        npcList.add(npc)
    }

    override fun startAsync() {
        executorService.scheduleWithFixedDelay({ this.performNpcMovementForAll() }, 0L, 100L, TimeUnit.MILLISECONDS)
    }

    override fun destroy() {
        try {
            executorService.shutdown()
        } catch (e: Exception) {
            log.error("Failed to shutdown executor service", e)
        }
        npcList.clear()
    }

    private fun performNpcMovementForAll() {
        npcList.forEach(npcMovementConsumer)
    }

    private fun performNpcMovement(npc: NpcModel) {
        if (npc.isMovementRunning) {
            return
        }
        if (System.currentTimeMillis() - npc.lastMovementMills < npc.repeatMovementInterval) {
            return
        }
        val taskResult = tiledMapCoreApi.performNpcMovement(npc.id, MovementDirection.randomDirection())
        taskResult?.run {
            npc.isMovementRunning = true
            thenRun {
                npc.isMovementRunning = false
                npc.lastMovementMills = System.currentTimeMillis()
            }
        }
    }
}